src/ophyd_async/epics/_backend/_p4p.py

```diff
@@ -39,6 +40,77 @@
 }
 
 
+def _data_key_from_value(
+    source: str,
+    value: Value,
+    *,
+    shape: Optional[list[int]] = None,
+    choices: Optional[list[str]] = None,
+    dtype: Optional[str] = None,
+) -> DataKey:
+    """
+    Args:
+        value (Value): Description of the the return type of a DB record
+        kwargs: Overrides for the returned values.
+            e.g. to force treating a value as an Enum by passing choices
+
+    Returns:
+        DataKey: A rich DataKey describing the DB record
+    """
+    shape = shape or []
+    dtype = dtype or specifier_to_dtype[value.type().aspy("value")]
+    display_data = getattr(value, "display", None)
+
+    d = DataKey(
+        source=source,
+        dtype=dtype,
+        shape=shape,
+    )
+    if display_data is not None:
+        for key in common_meta:
+            attr = getattr(display_data, key, nan)
+            if isinstance(attr, str) or not isnan(attr):
+                d[key] = attr
+
+    if choices is not None:
+        d["choices"] = choices
+
+    if limits := _limits_from_value(value):
+        d["limits"] = limits
+
+    return d
+
+
+def _limits_from_value(value: Value) -> Limits:
+    valueAlarm = getattr(value, "valueAlarm", None)
+    _empty_limit = LimitPair(low=None, high=None)
+
+    def get_control_or_display_limit(limit: str) -> LimitPair:
+        if (control_or_display := getattr(value, limit, None)) is None:
+            return _empty_limit
+        low = getattr(control_or_display, "limitLow", None)
+        high = getattr(control_or_display, "limitHigh", None)
+        return LimitPair(
+            low=None if isnan(low) else low, high=None if isnan(high) else high
+        )
+
+    def get_alarm_or_warning_limit(limit: str) -> LimitPair:
+        if valueAlarm is None:
+            return _empty_limit
+        low = getattr(valueAlarm, f"low{limit}Limit", None)
+        high = getattr(valueAlarm, f"high{limit}Limit", None)
+        return LimitPair(
+            low=None if isnan(low) else low, high=None if isnan(high) else high
+        )
+
+    return Limits(
+        alarm=get_alarm_or_warning_limit("Alarm"),
+        control=get_control_or_display_limit("control"),
+        display=get_control_or_display_limit("display"),
+        warning=get_alarm_or_warning_limit("Warning"),
+    )
```

Comment:
Could this be written more simply?

```suggestion
def _limits_from_value(value: Value) -> Limits:
    def get_limits(substucture_name: str, infix: str = "") -> LimitPair:
        substructure = getattr(value, substucture_name, None)
        low = getattr(substructure, f"low{infix}Limit", None)
        high = getattr(substructure, f"high{infix}Limit", None)
        return LimitPair(
            low=None if isnan(low) else low, high=None if isnan(high) else high
        )

    return Limits(
        alarm=get_limits("valueAlarm", infix="Alarm"),
        control=get_limits("control"),
        display=get_limits("display"),
        warning=get_limits("valueAlarm", infix="Warning"),
    )
```

Reply:
Not quite, it's low{infix}Limit but limitLow for the others.

Reply:
These can't be combined like this because it's lowLimit on one and limitInfixLow on the other.

Reply:
So I am inclined to leave it as it is.

Reply:
How about?
```suggestion
def _limits_from_value(value: Value) -> Limits:
    def get_limits(substucture_name: str, low_name: str = "limitLow", high_name: str = "limitHigh") -> LimitPair:
        substructure = getattr(value, substucture_name, None)
        low = getattr(substructure, low_name, None)
        high = getattr(substructure, high_name, None)
        return LimitPair(
            low=None if isnan(low) else low, high=None if isnan(high) else high
        )
    return Limits(
        alarm=get_limits("valueAlarm", "lowAlarmLimit", "highAlarmLimit"),
        control=get_limits("control"),
        display=get_limits("display"),
        warning=get_limits("valueAlarm", "lowWarningLimit", "highWarningLimit"),
    )
```

Reply:
Still no. Is this worth blocking on? Can i make an issue to put in the upcoming sprint to refactor it, it would be a good first issue.

Reply:
Got there in the end (assuming this passes), made a minor change to getattr(foo, bar, nan) as isnan(None) becomes unhappy. I'm assuming that if the attr is there (substructure is not None and hasattr(substructure, attr)) it will never be None